node/src/utils/blob_cache.rs

```diff
@@ -0,0 +1,167 @@
+//! A reference cache for items.
+
+use std::{
+    collections::HashMap,
+    hash::Hash,
+    sync::{Arc, Weak},
+};
+
+use datasize::DataSize;
+
+/// A cache of serialized items.
+///
+/// Maintains a collection of weak references and automatically purges them in configurable
+/// intervals.
+#[derive(DataSize, Debug)]
+pub struct BlobCache<I> {
+    /// The actual blob cache.
+    #[data_size(skip)]
+    items: HashMap<I, Weak<Vec<u8>>>,
+    /// Interval for garbage collection, will remove dead references on every n-th access.
+    garbage_collect_interval: u16,
+    /// Counts how many items have been added since the last garbage collect interval.
+    put_count: u16,
+}
+
+impl<I> BlobCache<I> {
+    /// Creates a new cache.
+    pub fn new(garbage_collect_interval: u16) -> Self {
+        Self {
+            items: HashMap::new(),
+            garbage_collect_interval,
+            put_count: 0,
+        }
+    }
+}
+
+impl<I> BlobCache<I>
+where
+    I: Hash + Eq,
+{
+    /// Stores an serialized item (blob) in the cache.
+    ///
+    /// At configurable intervals (see `garbage_collect_interval`), the entire cache will be checked
+    /// and dead references pruned.
+    pub fn put(&mut self, id: I, item: Weak<Vec<u8>>) {
+        self.items.insert(id, item);
+
+        if self.put_count >= self.garbage_collect_interval {
+            self.items.retain(|_, item| item.strong_count() > 0);
+
+            self.put_count = 0;
+        }
+
+        self.put_count += 1;
```

Comment:
Might be worth making an alias for `u16` and using that as the type for `garbage_collect_interval` and `put_count` to help avoid future refactoring allowing this to overflow?

Reply:
I do like that `u16` puts a reasonable limit on the garbage collection interval, though it might be too small.

That aside, anyone changing the type will get a type error due to line 48?

```
        if self.put_count >= self.garbage_collect_interval {
```

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=912b43e614b9bdb9d16c05203ad65107

Reply:
OK.  It's a nitpick, so I won't insist.  I knew there'd be a compiler error, but the knee-jerk reaction to fix the compiler error I think would be to cast the smaller type to the larger.  Using an alias for `u16` as the type for `put_count` and for `garbage_collect_interval` gives a slightly stringer hint as to the author's intent that these really should remain the same type.

Not a big deal though - I'm ok with it as is.